File: Sources/ExchangeLib/Exchange.swift
import Foundation
import Dispatch

public class Exchange
{
    var orderBooks : [String : OrderBook] = [:]
    var dispatchQueues : [String : DispatchQueue] = [:]
    var tradeQueue = DispatchQueue(label: "Trades")
    public var trades : [Trade] = []

    public init()
    {
    }

    public func insert(order: Order)
    {
        var orderBook = orderBooks[order.instrument]

        if orderBook == nil
        {
            orderBook = OrderBook()
            orderBooks[order.instrument] = orderBook
            dispatchQueues[order.instrument] = DispatchQueue(label: order.instrument)
        }

        if let dispatchQueue = dispatchQueues[order.instrument]
        {
            dispatchQueue.async
            {
                let trades = orderBook!.insert(order: order)

                self.tradeQueue.async
                {
                    self.trades += trades
                }
            }
        }
    }

    public func waitForTrades()
    {
        for dispatchQueue in dispatchQueues.values
        {
            dispatchQueue.sync {}
        }

        tradeQueue.sync {}
    }
}

File: part_1/go_solution/exchange.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
)

type Exchange struct {
	book map[string]*OrderBook
}

func NewExchange() *Exchange {
	return &Exchange{book: make(map[string]*OrderBook)}
}

func (ex *Exchange) Book(ins string) *OrderBook {
	if ob, ok := ex.book[ins]; ok {
		return ob
	}
	ob := &OrderBook{Instrument: ins}
	ex.book[ins] = ob
	return ob
}

func (ex *Exchange) Execute(o Order) ([]Trade, error) {
	ob := ex.Book(o.Instrument)
	ob.Insert(o)
	return ob.Match()
}

func main() {
	verbose := false
	exchange := NewExchange()
	reader := csv.NewReader(os.Stdin)
	reader.FieldsPerRecord = 4
	reader.Comma = ':'
	reader.ReuseRecord = true
	for {
		rec, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			panic(err)
		}
		o, err := ScanOrder(rec)
		if err != nil {
			panic(err)
		}

		if verbose {
			fmt.Println(o)
		}
		trades, err := exchange.Execute(*o)
		if err != nil {
			panic(err)
		}
		if len(trades) > 0 {
			for i := range trades {
				fmt.Println(trades[i])
			}
		}
		if verbose {
			fmt.Println(exchange.book[o.Instrument])
		}
	}
}
